\documentclass{article}
\usepackage{graphicx}
\usepackage[utf8]{inputenc}
\usepackage{float}
\usepackage{hyperref}
\usepackage{booktabs}
\usepackage[margin=1in]{geometry}
\usepackage{booktabs}
\usepackage{booktabs}
\usepackage{subcaption}
\linespread{1.5}
\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    filecolor=magenta,      
    urlcolor=cyan,
    citecolor=cyan
}



\urlstyle{same}


\begin{document}

\title{CoRNonCOB Application Note
	   \medskip \large}
\author{Ayesha Ahmad, Ethan Holleman}

\maketitle

\newpage

\section{Abstract}
CoRNonCOB (Compare Really Non-Conserved Oligopeptides in Bacteria) is a tool that identifies peptides in the non-coding regions of bacterial genomes that may contribute to a phenotype of interest. CoRNonCOB is designed to work at the very specific scale of individual phenotypes of specific strains of bacteria and because of this RNA-seq data that could illuminate the expression profiles of the bacteria of interest may not be available. Therefore, CoRNonCOB attempts to leverage genomic data by identifying peptides in non-coding regions that are conserved in the members of a positive phenotype and absent in a control.

With this tool, we hope to identify the peptides that are unique to \emph{Lactobacillus crispatus} that produce a microbial agent to kill E. coli bacteria. Through identifying these unique peptides, we hope to provide another parameter to characterize phenotypes. Our tool takes x amount of genomes in FASTA format  from each phenotype, and generates the peptides unique to each respective phenotype. Our tool is available at the \href{ https://github.com/EthanHolleman/CoRNonCOB}{GitHub page here}, where additional information on the parameters required and our current analyses can be found. Documentation is available at the \href{https://ethanholleman.github.io/CoRNonCOB/index.html}{CoRNonCOB website.}

\section{Introduction}
Currently, there are no tools for identifying peptides located in the non-coding regions of bacterial genomes.We are trying to identify peptides in the non-coding regions because for the \emph{L. crispatus} genomes that kill \emph{E. coli}, no antimicrobial peptides have been identified in their coding regions. Thus, we wanted to see if these peptides could be found in the non-coding regions. 


 However, there are tools that identify peptides in other organisms like plants. For example, the Small Peptide Alignment Discovery Application (SPADA), is an application to identify the small peptides in plant genomes. SPADA uses homology-based modeling to predict peptides that are one to two exons in length. Unfortunately, SPADA is not equipped to handle bacterial genomes because their genomes do not consist of introns \cite{zhou_silverstein_gao_walton_nallu_guhlin_young_2013}.


There is a database that stores the current known antimicrobial peptides called the Antimicrobial Peptide Database (APD). APD has aided in classifying Antimicrobial Peptides (AMP) and identifying peptides with similar sequences \cite{wang_2004}. AMPs are peptides associated with the ability to kill bacteria, viruses and fungi \cite{izadpanah}. CoRNonCOB is designed to identify peptides in the non-coding regions of bacterial genomes. For this specific project, we are looking at \emph{L. crispatus.} \emph{L. crispatus} are associated with a healthy vaginal and urinary tract microbiome environment \cite{becknell_spencer_2015}. In fact, Lactobacillus probiotics have been found to prevent recurrent Urinary Tract Infections (UTIs) in women \cite{dwyer_dwyer_2012}.


There are two phenotypes of \emph{L. crispatus} that we hope to distinguish by identifying unique peptides. One phenotype is characterized by their ability to produce a microbial agent lethal to \emph{E. coli}, while the other phenotype does not appear to produce any similarly lethal agent. With CoRNonCOB we hope to extract candidate AMPs that may help identify the true antimicrobial agent.


\section{Implementation}

\subsection{Workflow}
CoRNonCOB is written in python and incorporates Prokka for prokaryotic gene prediction, BioPython for sequence translation and IO, the CD-HIT suite of programs for sequence clustering and the modelAMP package for chemical property
calculations. The generalized workflow can be seen in figure \ref{fig:flow}.


\begin{figure}[h]
\centering
\includegraphics[width=12cm]{/home/ethan/Documents/github/CoRNonCOB/sphinx/images/diagram.png}
\caption{Histogram of candidate peptide properties identified from Putonti Lab \emph{Lactobacillus crispatus} genomes vs. all AMPs available from the \href{http://aps.unmc.edu/AP/main.php}{AMP3 Database}.}
\label{fig:flow}
\end{figure}
The user supplies paths to directories containing the genomes of the members of each phenotype. Running Prokka on each genome produces a gff file of predicted coding regions which CoRNonCOB then reads and inverts to yield predicted non-coding regions for each genome. Non-coding regions are translated into all six reading frames, written as fasta files in the program output. These fasta files representing the predicted non-coding peptides from individual genomes are then concatenated to create a file containing the predicted non-coding peptides of an entire phenotype. CD-HIT is then used to cluster the concatenated non-coding peptide files by sequence similarity. CD-HIT uses a greedy algorithm that sorts sequences by length and then searches for similar sequences via kmer comparison allowing for extremely fast sequence clustering \cite{10.1093/bioinformatics/btl158}. The longest sequence in each cluster is considered the representative sequence for that cluster. The user can control the minimum overlap of subject to representative sequence and minimum identify thresholds in the CoRNonCOB arguments. 


Peptide clusters meeting a certain threshold of genome participation are considered conserved and written to a new fasta file representing the predicted conserved non-coding peptides for an individual phenotype. CD-HIT-2D, which uses the same basic algorithm of CD-HIT but looks for peptides in one library that are similar to another library, compares the conserved non-coding peptides between the two phenotypes. CoRNonCOB then parses this output and removes non-unique sequences to produce a fasta file of candidate peptides that are uniquely conserved in the positive phenotype. Lastly, using the modelAMP python package basic chemical properties of each peptide are predicted and written to a csv file.

\subsection{Run Time and Memory Usage Estimates}


\begin{figure}[H]
  \centering
  \begin{minipage}[b]{0.45\textwidth}
    \includegraphics[width=\textwidth]{/home/ethan/Documents/github/CoRNonCOB/sphinx/images/memory.png}
    \caption{CoRNonCOB Memory usage with increasing numbers of genomes.}
    \label{fig:memory}
  \end{minipage}
  \hfill
  \begin{minipage}[b]{0.4\textwidth}
    \includegraphics[width=\textwidth]{/home/ethan/Documents/github/CoRNonCOB/sphinx/images/O.png}
    \caption{CoRNonCOB wall-clock run-time with increasing numbers of genomes.}
    \label{fig:runtime}
  \end{minipage}
\end{figure}

Figure \ref{fig:runtime} shows the time for CoRNonCOB to execute given an increasing number of genomes. All of the genomes used were of similar length. With this metric CoRNonCOB appears to run in linear time. Operations performed in python by CoRNonCOB such as non-coding region extraction and translation are done in linear time with respect to number of genomes and number of non-coding regions. 



However, Prokka is by far the rate limiting step in CoRNonCOBâ€™s run-time which is estimated as O(n\textsuperscript{2}) where n is the length of the genome used for prediction. Therefore if genome size remains relatively constant, run-time will reflect figure \ref{fig:runtime} but will increase to O(n\textsuperscript{2}) as genome size increases. Memory usage is estimated approximately O(n) where n is the number of genomes processed. The memory usage for the runs shown in figure \ref{fig:runtime} are plotted in figure \ref{fig:memory}.


\section{Results} 

In order to test the performance of CoRNonCOB we created a collection of 18 experimentally verified AMPs and randomly divided the peptides into positive or negative tests. Then, using three \emph{Lactobacillus crispatus} genomes provided by the Putonti lab and their collaborators randomly inserted the test peptides into these genomes. Three genomes were provides, two of the \emph{E. coli} lethal phenotype or "killer" and one wild-type with no \emph{E. coli} lethality, or "nice". Positive test peptides were inserted only into genomes with the killer phenotype and negative test peptides where inserted into all genomes regardless of phenotype. This is shown in the table below.


\begin{table}[H]
\centering
\begin{tabular}{@{}lllll@{}}
\toprule
\textbf{Test Type} &
  \textbf{Peptides in Test} &
  \textbf{\begin{tabular}[c]{@{}l@{}}Peptides Expected\\ in Final Results\end{tabular}} &
  \textbf{\begin{tabular}[c]{@{}l@{}}Peptides in\\ Final Results\end{tabular}} &
  \textbf{Percent Correctness} \\ \midrule
Positive &
  9 &
  9 &
  7 &
  77 \\
Negative &
  9 &
  0 &
  0 &
  100 \\ \bottomrule
\end{tabular}
\label{tab:prefom}
\end{table}

Currently, the CoRNonCOB pipeline has been able to recover seven of the nine positive peptides and filter out nine of the nine negative test peptides. Additionally our test run returned a total of 2293 candidate non-coding peptides unique to the positive (\emph{E. coli} lethal) phenotype. Figure \ref{fig:hist} shows histograms comparing properties of the candidates to all peptides in the AMP3 database. 


\begin{figure}[h]
\centering
\includegraphics[width=13cm]{/home/ethan/Documents/github/CoRNonCOB/sphinx/images/pep_props.png}
\caption{Histogram of candidate peptide properties identified from Putonti Lab \emph{Lactobacillus crispatus} genomes vs. all AMPs available from the \href{http://aps.unmc.edu/AP/main.php}{AMP3 Database}. Length is in amino acids. The Boman index is a measure of potential for protein-protein interaction \cite{boman_2003}. Hydrophobic ratio is ratio of hydrophobic amino acids over number of amino acids in the peptide.}
\label{fig:hist}
\end{figure}

Using the Putonti lab genomes CoRNonCOB identified a total of 4580 candidate peptides. The majority of candidate peptides  are less than 25 amino acids in length, with a median length of 18 residues.

\newpage
\bibliography{refs.bib}{}
\bibliographystyle{plain}

\end{document}