

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>io_utils module &mdash; CoRNonCOB 0.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="main module" href="main.html" />
    <link rel="prev" title="genome module" href="genome.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoRNonCOB
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Project Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="files.html">Output File Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="calander.html">Project Calander</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="args.html">args module</a></li>
<li class="toctree-l2"><a class="reference internal" href="genome.html">genome module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">io_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="main.html">main module</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype.html">phenotype module</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequence.html">sequence module</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html#object-structure">Object Structure</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contrib.html">Adding to this Site</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoRNonCOB</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="modules.html">Modules</a> &raquo;</li>
        
      <li>io_utils module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/io_utils.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-io_utils">
<span id="io-utils-module"></span><h1>io_utils module<a class="headerlink" href="#module-io_utils" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="io_utils.convert_genome_to_header_dict">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">convert_genome_to_header_dict</code><span class="sig-paren">(</span><em class="sig-param">genome_path</em>, <em class="sig-param">format='fasta'</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.convert_genome_to_header_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads in a file type (likely fasta) and creates a dictionary where keys
are the sequence headers and the values are lists of nucleotides. Each
index in the list will hold one nucleotide.
:param: genome_path: String. Path to the file where genomic sequences are stored
:param: format: String. File format. Default = fasta</p>
</dd></dl>

<dl class="function">
<dt id="io_utils.convert_genome_to_list">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">convert_genome_to_list</code><span class="sig-paren">(</span><em class="sig-param">genome_path</em>, <em class="sig-param">format='fasta'</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.convert_genome_to_list" title="Permalink to this definition">¶</a></dt>
<dd><p>DEPRECIATED</p>
<p>Takes in the path to a genome in the format specified by the format
variable. Genome file is read using SeqIO from Biopython and if the file
contains multible records the nucleotide sequences are appended to one
string. String is then returned as a list with each index containing one
nucleotide.</p>
</dd></dl>

<dl class="function">
<dt id="io_utils.filter_prokka_files">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">filter_prokka_files</code><span class="sig-paren">(</span><em class="sig-param">prokka_results_path</em>, <em class="sig-param">*args</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.filter_prokka_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Function for returning specific types of files from a directory where
prokka results have been written. Select the filetypes that should be
returned by including the file extensions without the . as arguements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prokka_results_path</strong> – String; Path to directory containing prokka results</p></li>
<li><p><strong>*args</strong> – <p>Strings; File extensions for the types of files to be returned</p>
</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="io_utils.if_not_exists_make">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">if_not_exists_make</code><span class="sig-paren">(</span><em class="sig-param">parent_dir_path</em>, <em class="sig-param">child_dir_name</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.if_not_exists_make" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if a child directory exists within a parent directory. If the child
directory does not exist makes it. Returns the path to the child directory
whether it was made or not. Does not search the parent directory
recursively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parent_dir_path</strong> – String; Path to the parent directory</p></li>
<li><p><strong>child_dir_name</strong> – String; Basename of directory to check if exists</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="io_utils.parse_cdhit_output_file">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">parse_cdhit_output_file</code><span class="sig-paren">(</span><em class="sig-param">clstr_file</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.parse_cdhit_output_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates through a cd-hit output file and returns a list of clusters.
The index of the cluster is equal to the cluster number. Each index contains
the sequences that cd-hit assigned to that cluster represented as a list
of tuples. Over data structure looks like [[()]]. Formating of the data
inside the tuples which represent the actual sequences is determined
by the parse_cdhit_record function.</p>
</dd></dl>

<dl class="function">
<dt id="io_utils.parse_cdhit_record">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">parse_cdhit_record</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">genome_id=True</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.parse_cdhit_record" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses a single record representing a peptide sequence in a cdhit output
file. Returns a tuple of that rows contents with extra characters and
white spaces cleaned up. Tuple contains the index of the sequence in the
cluster, the length of the sequence, the header of the sequence and the
alignment information in that order. If genome_id is True meaning a genome
id has been appended the end of each header this function will add that
on to the end of the tuple.</p>
<p>Currently assumes that cd-hit was run with -d 0 so full header is included
in the output and -p 1 so alignment statistics are printed to the output
file.</p>
</dd></dl>

<dl class="function">
<dt id="io_utils.parse_gff">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">parse_gff</code><span class="sig-paren">(</span><em class="sig-param">gff_path</em>, <em class="sig-param">*args</em>, <em class="sig-param">header=False</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.parse_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a path to a gff file returns as a list of lists the information
contained in the field indicies given via <a href="#id3"><span class="problematic" id="id4">*</span></a>args. All args should be
integers.</p>
<p>1   sequence  # subtract 1 from these!!!!!
2   source
3   feature
4   start
5   end
6   score
7   strand
8   phase
9   attributes</p>
</dd></dl>

<dl class="function">
<dt id="io_utils.read_phenotype_map">
<code class="sig-prename descclassname">io_utils.</code><code class="sig-name descname">read_phenotype_map</code><span class="sig-paren">(</span><em class="sig-param">map_path</em><span class="sig-paren">)</span><a class="headerlink" href="#io_utils.read_phenotype_map" title="Permalink to this definition">¶</a></dt>
<dd><p>If the user has genomes all over the place and does not
want to aggregate them they can provide a csv file with the
headers below to create Phenotype objects in the main
function.</p>
<p>PHENOTYPE, FILEPATH</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="main.html" class="btn btn-neutral float-right" title="main module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="genome.html" class="btn btn-neutral float-left" title="genome module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Ethan Holleman, Ayesha, Ahmad

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>